src/backends/perf_pt/mod.rs

```diff
@@ -260,10 +278,18 @@ impl Tracer for PerfPTTracer {
             return Err(HWTracerError::TracerAlreadyStarted);
         }
 
-        if unsafe { perf_pt_start_tracer(self.tracer_ctx) } == -1 {
+        // It is essential we box the trace now to stop it from moving. If it were to move, then
+        // the reference which we pass to C here would become invalid. The interface to
```

Comment:
Is this statement true? I assume you're not passing a poiunter to `PerfPTTrace`, at which point it doesn't matter if `PerfPTTrace` moves, so long as the C memory inside it doesn't?

Reply:
We are passing a pointer to a `PerfPTTrace`, as C needs to access the other fields too. 

Even if we were only passing the buffer itself, we would need to pass a `*mut *mut u8` which would have the same problem. Although the underlying C-level heap allocation won't move, the intermediate pointer might. The intermediate pointer is needed for a `realloc(3)` when we get around to implementing #33.

[I actually found this problem *before* i was passing a reference to the whole struct]

Reply:
OK, I think I get it.

Reply:
It's a mind bender, I agree. It took me most part of a day to find this bug!